### DISCUSSION TITLE
Switching to App router from Pages, is there any benefit when SSG is used

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#60935] [App Router] [answered]
Hello,
currently we have static site which is generated at build time. We use [[...slug]].js file in pages folder to catch all routes and export html at build time, then serve that over CDN. Our client needs are basically good SEO and performance which is pretty good using this approach.

However with new Nextjs features, app router, we are discussing would it be beneficial to switch to app router, use server components and client components, layouts etc ( if that makes sense as we are using SSG at the moment, and not SSR )

Does anyone have direct experience with this, or in other words would it be beneficial for performance, UX, DX and what not, to switch from pages to app using SSG approach, or to take advantages of app router we need to use SSR.

**Comment by joulev** [comment, marked as answer]
It depends a lot on what kind of application you currently have. For example I'd say blog apps will see a significant increase in performance, because the static blog content can "escape" hydration, while in the pages router they will always be hydrated as normal interactive React components.

The app router is good at both static rendering (SSG) and dynamic rendering (what you are referring to as SSR).

That said, if your app is running well in the pages router, I don't think you will see any significant gains by moving to the app router. So decide based on whether your team has more urgent deliverables to ship, whether your team is more comfortable maintaining a pages router or an app router application. Also play with the app router a bit and see if the team likes it.

Also don't forget that you can run two routers in parallel in the same app, so you can do incremental migration if needed.

**Reply by ghost** [reply]
It is marketing site app, with product listing, and lots of product details pages ( 200 + ) which all have static content that don't change that often, so we could say it is like classical blog app. Yes we have good performance but there is still room for improvements, as it always is.

Btw thanks for the answer, but can you please elaborate a bit more what you mean when you say "static blog content can "escape" hydration, while in the pages router they will always be hydrated as normal interactive React components."  

Does this means shiping less client side JS or what, every little bit counts, so I am open for migration to app router if there is benefit from it.

**Reply by joulev** [reply]
@mountinman In a nutshell you can think of it this way:

Your blog page will have a lot of HTML elements, say `<p>`. If these `<p>` are in client components, they always can have, say, custom `ref`, event handlers, interactivity added to it, hence React has to perform hydration to attach the event handlers correctly to these elements. If the `<p>` are in server components instead, React knows it is fully static, it's guaranteed to not have event handlers and interactivity, so React can skip hydration for such elements.

Hydration takes time and resource. For a typical React app, the number of elements being needlessly hydrated is not large so the performance impact is not significant. But say you make a blog post with lots of elements, with massive SVGs, with code blocks being server-side rendered to tons of syntax-highlighting elements using Shiki for example. Then the performance hit from hydration is no longer negligible.

The `app` router allows you to render things with server components (to skip hydration), but the `pages` router forces everything to be client components. Hence for mainly static websites, the `app` router will bring quite a significant gain in terms of hydration performance.

**Reply by ghost** [reply]
It is clear now, thanks.

**Reply by ankittchaurasia** [reply]
@joulev  You are totally wrong if you think nextjs don't hydrate server html. Nextjs has to hydrate every html because that's how React work.  check your source code, there will be content inside script tags even for server components.
See this discussion: #42170 
@mountinman There's no performance increase in app router for SSG but dev experience has increased significantly.
If it's marketing site with less js then better go with Astro
